gap/grahom.gi

```diff
@@ -652,6 +652,200 @@ function(src, ran, x, cols1, cols2)
     and DigraphsRespectsColouring(src, ran, x, cols1, cols2);
 end);
 
+InstallMethod(IsDigraphFolding, "for a digraph, a digraph, and a transformation",
+[IsDigraph, IsDigraph, IsTransformation, IsTransformation],
+function(domain, codomain, vertexmap, edgemap)
+  local temp, badedgepairs, optiontree, newoptions, currentpath,
+        goodconfigurations, newoptionstemp, codomedges, domedges,
+        domedgesbyinvertex, domedgesbyoutvertex, edgepath,
+        configurationstocheck, currentconfiguration, i, goodedgepairs,
+        newpair, codomedgesbyinvertex, newconfiguration, currentpair,
+        failedpairs, coverededges, loopable, forwardcoverededges;
+
+  #checking functions have the correct domain and codomain 
+  for i in [DigraphNrVertices(domain) + 1 ..
+            DegreeOfTransformation(vertexmap)] do
+    if not i ^ vertexmap = i then
+      return false;
+    fi;
+  od;
+  for i in [DigraphNrEdges(domain) + 1 .. DegreeOfTransformation(edgemap)] do
+    if not i ^ edgemap = i then
+      return false;
+    fi;
+  od;
+  if not IsSubset([1 .. DigraphNrVertices(codomain)],
+                  ImageSetOfTransformation(vertexmap,
+                                           DigraphNrVertices(domain))) then
+    return false;
+  fi;
+  if not IsSubset([1 .. DigraphNrEdges(codomain)],
+                  ImageSetOfTransformation(edgemap, DigraphNrEdges(domain))) then
+    return false;
+  fi;
+
+  #checking that the function is a multidigraph homomorphism 
+  for i in [1 .. DigraphNrEdges(domain)] do
+    temp := DigraphEdges(domain)[i];
+    if not DigraphEdges(codomain)[i ^ edgemap] =
+           [temp[1] ^ vertexmap, temp[2] ^ vertexmap] then
+      return false;
+    fi;
+  od;
+
+  domedges := DigraphEdges(domain);
+  codomedges := DigraphEdges(codomain);
+
+  domedgesbyinvertex := List(DigraphVertices(domain),
+                             x -> Filtered([1 .. Size(domedges)],
+                                           y -> domedges[y][1] = x));
+  codomedgesbyinvertex := List(DigraphVertices(codomain),
+                               x -> Filtered([1 .. Size(codomedges)],
+                                             y -> codomedges[y][1] = x));
+  domedgesbyoutvertex := List(DigraphVertices(domain),
+                              x -> Filtered([1 .. Size(domedges)],
+                                            y -> domedges[y][2] = x));
```

Comment:
It would be more efficient to loop through the edges than filter them, although this probably won't make much difference to the overall efficiency.